### ISSUE TITLE
Do Futures need to be Send?

### ISSUE THREAD

**Issue by bendk** [mozilla/uniffi-rs#1538]
In the async-refactor branch, I added the requirement that all async functions need to return Futures that are `Send`, with the following reasoning:

> The future needs to be `Send`, since it will move to whatever thread the foreign executor chooses.  However, it doesn't need to be `Sync', since we don't share references between threads (see do_wake()).

@ijc [reports that this is breaking their build](https://github.com/mozilla/uniffi-rs/pull/1515#discussion_r1192330185).  It looks like they might be always using the library in a single-threaded manner and they don't need or want the `Send` bound.

AFAICT, this is definitely possible to do and seems like a valid use case.  For example if you're only targeted Python with it's single-threaded `asyncio.EventLoop`.  Another example would be JS if we added support.  I don't believe there's an easy way to do this on Kotlin/Swift, but it seems possible for someone to write a custom code that implements it.

Is there a way we can support this use case?  I'm not sure exactly how it would be configured since the bound is in [uniffi_core itself](https://github.com/mozilla/uniffi-rs/blob/main/uniffi_core/src/ffi/rustfuture.rs#L146-L159).  Maybe we could use a feature flag.  Or maybe remove that bound and replace it with a bound that the scaffolding code generates somehow.

Note: I believe the same issue also applies to callback interfaces.

**Comment by bendk** [comment]
This could even apply to interfaces.  We could have a single-threaded mode where we don't require a `Send` bound and we use `Rc` instead of `Arc` for interfaces.

**Comment by mhammond** [comment]
How would that be enforced though? Isn't single threaded mode really just saying "the foreign code promises it will only use uniffi from exactly 1 thread" - would we still expect the Rust side of the world to detect violation of this? Or would not keeping the promise mean crashes are expected and part of the contract?

**Comment by bendk** [comment]
> Isn't single threaded mode really just saying "the foreign code promises it will only use uniffi from exactly 1 thread"

Good question.  I guess that's exactly what I'm proposing and when I read that I realize it's very restrictive.  I think it does makes sense for a library that wants to support JS consumers, but that might be the only main use case.

**Comment by mhammond** [comment]
> makes sense for a library that wants to support JS consumers

Yeah - if we can work out a way to "enforce" these assumptions there could be some nice wins.

**Comment by ijc** [comment]
I think I've resolved most of the cases I was seeing by adding the appropriate bounds etc. 

What remains points me towards a slightly looser requirement than I thought I had at first (and which got captured here, thanks Ben!). I'm also starting to doubt that this is a Uniffi problem.

I think the actual requirement I have is: Rust code which is written to be compatible with Uniffi's Foreign Executor should also be able to be compatible with non-Uniffi/native-Rust executors, including those which are single threaded i.e. those which require `Futures` to be `!Send`.

(Aside, I do think the requirement captured here is potentially interesting, it just isn't one I think I have right now).

However I'm starting to think that this is actually an issue with our use of async traits (via the [async-trait](https://crates.io/crates/async-trait) crate) and not Uniffi.

Some background on our use:

We have a Rust library which is embedded into our Android and iOS apps via Uniffi[^0], browser via [wasm-bindgen](https://crates.io/crates/wasm-bindgen/) and native Rust code (a CLI and Rust level integration tests).

Part of our API involves the application passing in a set of callbacks which we model as an async[^1] trait using [async-trait](https://crates.io/crates/async-trait).

Previously the async traits were `#[async_trait(?Send)]`. As part of the Uniffi bump in #1515 we had to switch this to `#[async_trait]`. That us what broke our JS bindings and our integration tests.

The JS side was actually fairly easy to deal with:
```rust
#[cfg_attr(not(target_arch = "wasm32"), async_trait)]
#[cfg_attr(target_arch = "wasm32", async_trait(?Send))]
```

But the integration tests have turned out to be less tractable. These are based on the Actix web test framework (since our server side component uses Actix) and Actix is fundamentally designed with all Futures as `!Send` and so the tests run on Tokio single threaded.

So I think this turns out to be actually more an issue with async traits than Uniffi -- the essential problem is that the `Send`/`!Send` has to be part of the trait type and cannot be inferred by the compiler as it would be for a normal Future[^2], which means the requirement can't be deferred until the Future comes into contact with the actual executor at which point either things work or the executor needs `Send` and the compiler tells us if any Future is `!Send`.

So far I don't have a way forward other than to make Uniffi support a cargo feature which is mutually incompatible with our integration tests and set the `?Send` on the trait as needed . I think that's not the absolute end of the world.

[^0]: Actually, Uniffi is currently just a PoC our current solution uses [flapigen](https://crates.io/crates/flapigen)
[^1]: In the Uniffi case we don't actually pass the async-ness over the FFI we use a synchronous `callback interface` and provide a shim over that to make it async. This happens to work for Swift and Kotlin since they have mechanisms for firing of async work from sync functions.
[^2]: I've not dug into the nightly async traits feature, but I suspect this must be case there too since `Send` it has to be part of the `dyn Foo` trait signature somewhere.

**Comment by bendk** [comment]
I was feeling like this was very messy, but the more I think about it a "single-threaded UniFFI" makes sense -- especially as we add more async features.  What if we defined UniFFI feature named `single-threaded`.  When enabled it would:
  - Remove the `Send` bound from Futures returned by async functions.
  - Remove the `Send + Sync` bound on interfaces
  - Generate thread-unsafe libraries.  Library authors would be responsible for informing their consumers of this.

The use case is exactly the one you bring up: libraries that run in single-threaded executors like Actix, Python, and JS.

One hesitation I have is there's another definition of "thread-unsafe": objects that can be shared between threads as long as access to those objects is serialized.  Kind-of like `Send + !Sync` in Rust terms.  If we want to support this, then it makes this work more complex and probably means we can't use a feature flag.  But I'm thinking we don't need to support this because I can't think of any good motivation.  Can others?

**Comment by bendk** [comment]
BTW, I'm hoping to implement async callbacks soon.  This would work very similar to `async-trait` -- you wrap your trait with a UniFFI attribute and it would convert `async fn foo() -> Bar` to `fn foo() -> Future<Output = Bar> + Send`.  If we implemented that and also made the `single-threaded` flag remove the `Send` bound, would that help your use case?

**Comment by mhammond** [comment]
> libraries that run in single-threaded executors like Actix, Python, and JS.

While Python *usually* just uses a single thread with asyncio, https://docs.python.org/3/library/asyncio-eventloop.html does have quite a few docs for how to do multi-threading - eg, which each thread has a single event-loop, it documents how multiple threads each with its own event loop works. Other functions are described as thread-safe or not. IIUC a JS implementation targeting node couldn't assume single-threaded. In both those cases it could *arrange* to be single-threaded, but so could non-async programs.

So while Actix and JS-in-Firefox look like they *must* be single-threaded, so must certainly be safe in those environments, I'm still slightly skeptical about this without a compelling reason and/or without some enforcement

**Comment by ijc** [comment]
Earlier I said:
> So far I don't have a way forward other than to make Uniffi support a cargo feature which is mutually incompatible with our integration tests and set the ?Send on the trait as needed . I think that's not the absolute end of the world.

I did implement that and it works, when running integration tests we just don't enable Uniffi bindings and when we build for deployment to Android/iOS we pass the feature flag to turn them on as part of our xcode/gradle project integration.

@bendk said:
> BTW, I'm hoping to implement async callbacks soon. This would work very similar to async-trait -- you wrap your trait with a UniFFI attribute and it would convert `async fn foo() -> Bar` to `fn foo() -> Future<Output = Bar> + Send`. If we implemented that and also made the `single-threaded` flag remove the Send bound, would that help your use case?

We want and like the multithreaded stuff when we are actually embedded into the Swift & Kotlin apps. The cases where we want Uniffi are mutually exclusive with the cases where we are required to support single threaded execution.

So _maybe_ we could use the uniffi macro as a drop in replacement for `async-trait`[^0] even in the cases when we aren't actually using Uniffi, but it might just make things more complex rather than easier?

The feature I've added to my library is additive, enabling it turns on Uniffi bindings, which means that e.g. `cargo test` just works (it uses Actix but not Uniffi). The proposed `uniffi/single-threaded` feature is in some way subtractive, it removes support for multi-threading. I think it might be tricky to work with a negative feature in general but especially for integration tests?

Maybe I should point out that our bindings tests do not use Actix, they run as native Android and iOS tests using a different doubling mechanism. The "integration" tests I talk about earlier are testing the core Rust functionality against the server with the database etc.

[^0]: I think that's the proposal rather than something generated into the scaffolding code which might live alongside `async-trait`?

**Comment by extraymond** [comment]
Once example I think non-send future would make things clearer is for having non-send futures passed as an interface or callback into rust.

I've learned that when trying to call an python callback from rust which assume to be on the same thread would segfault if my rust async runtime try to move it around. Just to clarify, this is totally avoidable with existing code by not spawning the future on another thread.

Suppose I'm creating an framework which accepts python code to implement certain trait and can be passed into rust, it would be nicer to maintain it if I can enforce that non-send nature on my rust code.
